### PR TITLE
#3914: gapi load error telemetry

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Part of #3914 

## Future Work

- This is an attempt to gather more info, this won't fix the issue (obviously)

## Checklist

- [ ] Add tests
- [ ] Designate a primary reviewer

### CODE REVIEW

File: src/contrib/google/initGoogle.ts [hunk 1]
```diff
@@ -56,6 +58,9 @@ function initGoogle(): void {
 
   const script = document.createElement("script");
   script.src = "https://apis.google.com/js/client.js?onload=onGAPILoad";
+  script.addEventListener("error", (error) => {
```
Comment (twschiller) [inline src/contrib/google/initGoogle.ts:61]: Should this be `script.onerror`? See here: https://developer.mozilla.org/en-US/docs/Web/API/HTMLScriptElement#dynamically_importing_scripts

Also, I don't think the argument is the error? I think it's the event?

Comment (BLoe) [reply src/contrib/google/initGoogle.ts]: Yep, sorry, was adding the onerror next, I should have waited to request review. Thanks for the catch on the argument though, it's the event not the error. Does it hurt to do both? I can remove this if you don't think we need it.

Comment (twschiller) [reply src/contrib/google/initGoogle.ts]: > Does it hurt to do both?

I don't think `script.addEventListener("error")` does anything? We should avoid adding code if we think that doesn't do anything

Comment (twschiller) [reply src/contrib/google/initGoogle.ts]: Nevermind, it looks like error works on individual elements too: https://developer.mozilla.org/en-US/docs/Web/API/Element/error_event

I'm not sure what the difference is in practice with onerror

We could leave both?

cc @fregante

File: src/contrib/google/initGoogle.ts [hunk 2]
```diff
@@ -53,9 +60,14 @@ function initGoogle(): void {
   }
 
   window.onGAPILoad = onGAPILoad;
+  window.onGAPIError = onGAPIError;
 
   const script = document.createElement("script");
-  script.src = "https://apis.google.com/js/client.js?onload=onGAPILoad";
+  script.src =
+    "https://apis.google.com/js/client.js?onload=onGAPILoad&onerror=onGAPIError";
```
Comment (BLoe) [inline src/contrib/google/initGoogle.ts:67]: @twschiller I don't know exactly how to test this, but I ran a Sheets extension in page editor after this change and it works fine.

My reasoning for this is that I've seen this pattern in most of the results when I was googling around for info on this:
```
<script async defer src="https://apis.google.com/js/api.js"
	onload="window.gapiLoaded()"
	onerror="window.gapiLoadError(event)"
>
```

My thought was that if you can tack the `onload` onto the end of the URL like we're already doing, then `onerror` should work the same way... Does that make sense? Kind of flimsy reasoning, idk.

Comment (BLoe) [reply src/contrib/google/initGoogle.ts:67]: I could also just attach both to the `script` object manually with `script.onload = ...` and `script.onerror = ...`

Comment (BLoe) [reply src/contrib/google/initGoogle.ts:67]: Whoops, never mind, we actually have a lint rule to prefer `addEventListener()` over the shorthand ones (onload, onerror, etc)

File: src/contrib/google/initGoogle.ts (hunk 1 above)
Comment (BLoe) [reply src/contrib/google/initGoogle.ts]: Well, if I remove the `?onload=` in the script src url and instead add an event handler to the script, it no longer works properly:
```
script.src = "https://apis.google.com/js/client.js";
  script.addEventListener("load", async () => {
    await window.onGAPILoad();
  });
```

This fires before the script finishes loading, which causes an NPE on the `gapi` object that the load handler references.

So it seems like putting it in the URL works differently than a regular `addEventListener` call.

File: src/contrib/google/initGoogle.ts (hunk 2 above)
Comment (twschiller) [reply src/contrib/google/initGoogle.ts:67]: > I don't know exactly how to test this

You would need to test it by blocking calls to `https://apis.google.com/js/client.js` I think (e.g., by changing the CSP or using an extension to block it)

The `onGAPILoad` that was in the code is the necessary code for the load event (it's called when gapi finished initializing). I'm not sure if onGAPIError exists (I don't see anything about it in Google). It's probably fine to keep. It's unlikely it does anything though

**Review by twschiller — APPROVED** [review]
I removed onGAPIError because there's no references to it in the `https://apis.google.com/js/client.js` source code

File: src/contrib/google/initGoogle.ts [hunk 3]
```diff
@@ -55,7 +63,12 @@ function initGoogle(): void {
   window.onGAPILoad = onGAPILoad;
 
   const script = document.createElement("script");
+  // NOTE: there's no onGAPIError handler. You can verify by reading the https://apis.google.com/js/client.js code,
+  //   you can search for onGAPILoad handler and you'll see it's there but there's no onGAPIError or similar handler
   script.src = "https://apis.google.com/js/client.js?onload=onGAPILoad";
+  script.addEventListener("error", (event) => {
+    reportError(event.error);
```
Comment (fregante) [inline src/contrib/google/initGoogle.ts:70]: This should already happen thanks to the global event handler, right? All errors should be included but I'm not 101% sure

Comment (twschiller) [reply src/contrib/google/initGoogle.ts:70]: I would think so, we're kind of just grasping at straws on this one and adding in telemetry in places

**Pull request merged**